lib/model/unreads.dart

```diff
@@ -0,0 +1,411 @@
+              }
+              switch (detail.type) {
+                case MessageType.stream:
+                  final perStream = (newlyUnreadInStreams[detail.streamId!] ??= {});
+                  final perTopic = (perStream[detail.topic!] ??= QueueList());
+                  perTopic.add(messageId);
+                case MessageType.private:
+                  final narrow = DmNarrow.ofUpdateMessageFlagsMessageDetail(selfUserId: selfUserId,
+                    detail);
+                  (newlyUnreadInDms[narrow] ??= QueueList())
+                    .add(messageId);
+              }
+            }
+            newlyUnreadInStreams.forEach((incomingStreamId, incomingPerTopic) {
+              incomingPerTopic.forEach((incomingTopic, incomingMessageIds) {
+                _addAllInStreamTopic(incomingMessageIds..sort(), incomingStreamId, incomingTopic);
+              });
+            });
+            newlyUnreadInDms.forEach((incomingDmNarrow, incomingMessageIds) {
+              _addAllInDm(incomingMessageIds..sort(), incomingDmNarrow);
+            });
+        }
+    }
+    notifyListeners();
+  }
+
+  // TODO use efficient lookups
+  bool _slowIsPresentInStreams(int messageId) {
+    return streams.values.any(
+      (perTopic) => perTopic.values.any(
+        (messageIds) => messageIds.contains(messageId),
+      ),
+    );
+  }
+
+  void _addLastInStreamTopic(int messageId, int streamId, String topic) {
+    ((streams[streamId] ??= {})[topic] ??= QueueList()).addLast(messageId);
+  }
+
+  // [messageIds] must be sorted ascending and without duplicates.
+  void _addAllInStreamTopic(QueueList<int> messageIds, int streamId, String topic) {
+    final perTopic = streams[streamId] ??= {};
+    perTopic.update(topic,
+      ifAbsent: () => messageIds,
+      // setUnion dedupes existing and incoming unread IDs,
+      // so we tolerate zulip/zulip#22164, fixed in 6.0
+      // TODO(server-6) remove 6.0 comment
+      (existing) => setUnion(existing, messageIds),
+    );
+  }
+
+  // TODO use efficient model lookups
+  void _slowRemoveAllInStreams(Set<int> idsToRemove) {
+    final newlyEmptyStreams = [];
+    streams.forEach((streamId, perTopic) {
+      final newlyEmptyTopics = [];
+      perTopic.forEach((topic, messageIds) {
+        messageIds.removeWhere((id) => idsToRemove.contains(id));
+        if (messageIds.isEmpty) {
+          newlyEmptyTopics.add(topic);
+        }
+      });
+      for (final topic in newlyEmptyTopics) {
+        perTopic.remove(topic);
+      }
+      if (perTopic.isEmpty) {
+        newlyEmptyStreams.add(streamId);
+      }
+    });
+    for (final streamId in newlyEmptyStreams) {
+      streams.remove(streamId);
+    }
+  }
+
+  void _removeAllInStreamTopic(Set<int> incomingMessageIds, int streamId, String topic) {
+    final perTopic = streams[streamId];
+    if (perTopic == null) return;
+    final messageIds = perTopic[topic];
+    if (messageIds == null) return;
```

Comment:
I think calling the intermediate variable here `topics` (like you used in the previous revision) instead of `perTopic` would work well. It's parallel to `streams` for the outer data structure, and `messageIds` for the inner.